Add runtime deps tracker interface and related definitions

pmvald opened the pull request:
The types and interfaces related to the runtime deps tracker are added. Also an empty implementation is added to be completed in follow up PRs (after the interfaces are confirmed in this PR).

The added files are not used anywhere, so the change should not affect anything in anyway.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

pmvald committed:
refactor(core): add runtime deps tracker public interface with an empty implementation

The types and interfaces re;ated to the runtime deps tracker are added. Also an empty implementation is added to be completed in follow up PRs (after the interfaces are confirmed in this PR).

The added files are not used anywhere, so the change should not affect anything in anyway.

pkozlowski-opensource commented:
@pmvald looks like it will require rebase after recent merges.

pmvald commented:
> @pmvald looks like it will require rebase after recent merges.

Done!

devversion commented on packages/core/src/render3/deps_tracker/api.ts:
```diff
@@ -0,0 +1,66 @@
+export interface DepsTrackerApi {
+  /**
+   * Computes the component dependencies, i.e., a set of components/directive/pipes that could be
+   * present in the component's template (This set might contain elements not present in the
```
```suggestion
   * present in the component's template (This set might contain directives/components not necessarily used in the
```

pmvald committed:
refactor(core): tweak runtime deps tracker to accommodate JIT requirements.

The previous API design for the runtime deps tracker was done with only local compilation in mind. Now if we want to use it for JIT as well then some new requirements come along which are addressed by this commit as follows:
  - JIT scopes need distinct sets for directive and pipes. Thus the ScopeData interface is modified to include these info.
  - JIT needs access to scopes for various operations. So methods `getNgModuleScope` and `getStandaloneComponentScope` are added to the tracker's public api.
  - Test bed needs to clear the scope cache for its override mechanism. So method `clearScopeCacheFor` is added to the tracker's public api for this reason.

pmvald commented:
@devversion If we want to use this api for JIT as well then we need some more stuff. I added a new commit to this PR to address the JIT specific needs. See if it makes sense. The previous commit is as it is.

devversion reviewed (approved):
Seems reasonable to me. LGTM. I think we agreed on using the deps tracker for JIT as well

dylhunn reviewed (approved):
reviewed-for: fw-core

AndrewKushnir commented:
This PR was merged into the repository by commit 7d7fc48e58f51e574f3a5c1174cd7d95b515ddc8.
